nav2_costmap_2d/plugins/costmap_filters/speed_filter.cpp

```diff
@@ -0,0 +1,349 @@
+ *
+ * Author: Alexey Merzlyakov
+ *********************************************************************/
+
+#include "nav2_costmap_2d/costmap_filters/speed_filter.hpp"
+
+#include <cmath>
+
+#include "tf2/convert.h"
+#include "tf2_geometry_msgs/tf2_geometry_msgs.h"
+
+#include "nav2_costmap_2d/costmap_filters/filter_values.hpp"
+
+namespace nav2_costmap_2d
+{
+
+SpeedFilter::SpeedFilter()
+: filter_info_sub_(nullptr), mask_sub_(nullptr),
+  speed_limit_pub_(nullptr), filter_mask_(nullptr), mask_frame_(""), global_frame_(""),
+  speed_limit_(NO_SPEED_LIMIT), speed_limit_prev_(NO_SPEED_LIMIT)
+{
+}
+
+void SpeedFilter::initializeFilter(
+  const std::string & filter_info_topic)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  // Declare "speed_limit_topic" parameter specific to SpeedFilter only
+  std::string speed_limit_topic;
+  declareParameter("speed_limit_topic", rclcpp::ParameterValue("speed_limit"));
+  node->get_parameter(name_ + "." + "speed_limit_topic", speed_limit_topic);
+
+  filter_info_topic_ = filter_info_topic;
+  // Setting new costmap filter info subscriber
+  RCLCPP_INFO(
+    logger_,
+    "SpeedFilter: Subscribing to \"%s\" topic for filter info...",
+    filter_info_topic_.c_str());
+  filter_info_sub_ = node->create_subscription<nav2_msgs::msg::CostmapFilterInfo>(
+    filter_info_topic_, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&SpeedFilter::filterInfoCallback, this, std::placeholders::_1));
+
+  // Get global frame required for speed limit publisher
+  global_frame_ = layered_costmap_->getGlobalFrameID();
+
+  // Create new speed limit publisher
+  speed_limit_pub_ = node->create_publisher<nav2_msgs::msg::SpeedLimit>(
+    speed_limit_topic, rclcpp::QoS(10));
+  speed_limit_pub_->on_activate();
+
+  // Reset speed conversion states
+  base_ = BASE_DEFAULT;
+  multiplier_ = MULTIPLIER_DEFAULT;
+}
+
+void SpeedFilter::filterInfoCallback(
+  const nav2_msgs::msg::CostmapFilterInfo::SharedPtr msg)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (!mask_sub_) {
+    RCLCPP_INFO(
+      logger_,
+      "SpeedFilter: Received filter info from %s topic.", filter_info_topic_.c_str());
+  } else {
+    RCLCPP_WARN(
+      logger_,
+      "SpeedFilter: New costmap filter info arrived from %s topic. Updating old filter info.",
+      filter_info_topic_.c_str());
+    // Resetting previous subscriber each time when new costmap filter information arrives
+    mask_sub_.reset();
+  }
+
+  // Set base_/multiplier_ or use speed limit in % of maximum speed
+  base_ = msg->base;
+  multiplier_ = msg->multiplier;
+  if (msg->type == SPEED_FILTER_PERCENT) {
```

Comment:
Curious- how does the info publisher fill in this type for users? This is the same type of question we had from keepout PR to make sure that this work can be actually used on merging with tools ready.

Reply:
https://github.com/ros-planning/navigation2/blob/main/nav2_map_server/src/costmap_filter_info/costmap_filter_info_server.cpp#L31 

Ah we have a param for that, just make sure to document that this needs to be changed to certain values in docs.

Reply:
Currently has correct description starting from https://github.com/ros-planning/navigation.ros.org/pull/122/commits/e178508382727e5ab0cab3f631d8b5960d941259#diff-587042f9785caf9e27687ed0a0cd4fc919fd2d05687e7247145433ec36a59b10R100

Reply:
Also, back-ported an CostmapFilterInfo update from official documentation to current repos' docs: https://github.com/ros-planning/navigation2/pull/2074/commits/943d052053af12f4f5cf1fdb15715b0e7b0911de